Add solution to ex06 of UML, using collections (lists)

PmaiWoW opened the pull request.

PmaiWoW committed:
Add solution to ex06 of UML, using collections (lists)

nunofachada reviewed (changes requested):
Olá Pedro,

* Enunciado mal copiado.
* Para que serve a lista de jogadores em `Program`? Repara que no UML não há relação direta entre `Program` e `Player`, pelo que essa lista vai contra o diagrama.
* Porquê estar a instanciar a classe `Program` se a única coisa que o respetivo método `Run` faz é imediatamente instanciar outra classe? Parece-me perfeccionismo a mais só para teres um `Main` com duas linhas em vez de três. Além disso, no UML a classe `Program` não tem um método `Run`.
* No teu código a classe `Game` não contém nenhum `Map` (composição), apenas faz uso de `Map`s (ou seja, o código apenas apresenta uma dependência).

Para já fico por aqui. Ajusta o código segundo estes comentários e vê se é preciso mudares mais alguma coisa relativamente ao resto.

PmaiWoW committed:
Update solution to ex06 of UML

PmaiWoW committed:
Fix question in solution file of ex06 of UML

PmaiWoW committed:
Remove unwanted line

nunofachada reviewed (changes requested):
Já te digo mais alguma coisa sobre este exercício, mas entretanto corrige o _link_ para a imagem no enunciado da solução.

nunofachada reviewed (changes requested):
* Continuo sem perceber para que serve a lista que tens em `Program`.
* Nas próximas submissões testa e corre o código em Visual Studio, confirma que não há erros e _warnings_. Por exemplo, essa tal lista na classe `Program` deve dar um _warning_ de não ser usada.
* Qual a visibilidade dos diferentes mapas que estão como variáveis de instância em `Game`? Atenção que `Game` não tem grandes relações com `OpenWorldMap` e `DungeonMap`, apenas com `Map`. Lembra-te, é sempre preferível programar para uma _interface_ (ou o que for mais geral ou abstrato possível), do que para classes concretas. Obviamente que em alguma altura terás de instanciar mapas concretos, mas podes fazê-lo apenas no método `GameLoop()`, e nesse caso terias apenas uma dependência entre `Game` e esses mapas concretos (e muitas vezes algumas dependências omitem-se do diagrama, pelo que dessa forma não estarias a ir contra o diagrama).
* Classe `DungeonMap`: atenção que as variáveis locais começam com minúscula; parâmetros de métodos são variáveis locais.
* Classe `Player`: será que faz sentido `health` e `shield` serem `readonly`? Num jogo esses valores nunca mudam?

PmaiWoW committed:
Update solution to ex06 of UML

nunofachada reviewed (changes requested):
* O _link_ para a imagem do enunciado não foi ainda corrigido.
* Qual a visibilidade das variáveis de instância do tipo `Map` na classe `Game`?
* Uns comentários dentro do método `GameLoop()` não faziam mal nenhum.
* Onde está o método `PrintPlayerNames()` da classe `Map`?
* Nos métodos `ToString()` de `OpenWorldMap` e `DungeonMap` porque não aproveitas a funcionalidade já oferecida pela classe `Map`? É que estás a repetir código.
* Na classe `Player`: não é boa prática usar o operador Lambda em métodos que devolvem `void`.

PmaiWoW committed:
Update solution to ex06 of UML

PmaiWoW commented:
O link no enunciado vai dar à imagem certa, mas na minha pré-visualização a imagem em si não aparece, por alguma razão, ainda estou a tentar descobrir o porquê disto, por enquanto fiz push das alterações feitas até agora.

nunofachada reviewed (changes requested):
* A imagem aparece bem agora.
* Na classe `Map`, a variável `gamesPlayed` não é de instância, pois tem um modificador específico que está indicado no UML mas que neste momento não está no código.
* Qual a visibilidade das variáveis de instância do tipo `Map` na classe `Game`? (já te ando a perguntar isto há algum tempo...)
* A relação entre `Map` e `Player` não é de composição. Logo os jogadores não devem ser criados pelo mapa / dentro da classe `Map`.

PmaiWoW committed:
Update solution to ex06 of UML

Remove using directives that were unnecessary to certain classes;
Creation of the generic collection of Player objetcs (list) is now done in
the Game class instead of the Map class.

nunofachada reviewed (changes requested):
* Na classe `Game`, a variável `gamesPlayed` está sublinhada. Isso não é reflectido no código.
* Não existe relação direta entre `Game` e `Player`, ou seja, no máximo poderei aceitar uma dependência (que são por vezes omitidas para simplificar os diagramas). Nunca uma composição, que é o que  tens agora.

PmaiWoW commented:
Estou um pouco confuso onde poderia criar/ter a lista dos jogadores, considerando que não poderá estar na classe `Map` ou `Game`, então.
Poderiam estar lá se a lista fosse declarada dentro do main, não tendo _scope_ de classe mas apenas de método?

nunofachada commented:
Yes. Dependência.

PmaiWoW committed:
Update solution to ex06 of UML

nunofachada reviewed (changes requested):
Gostava mais do último `Console.WriteLine()` no sítio onde ele estava (no método `Main()`), pois isso realçaria o caráter `static` da variável `gamesPlayed`.

PmaiWoW committed:
Update solution to ex06 of UML

nunofachada reviewed (changes requested):
* O diagrama UML diz que um `Game` tem apenas um `Map`. No entanto o teu `Game` tem quatro mapas (ou melhor, quatro variáveis do tipo `Map`).
* Estás sempre a passar a mesma lista de jogadores a cada mapa, adicionando jogadores a cada novo mapa. Será que isso faz sentido, ou faria mais sentido cada mapa que instancias ter a sua própria lista de jogadores?
* Na classe `Map` e suas derivadas, é necessário passares explicitamente uma `List<Player>`? Não podes passar algo mais abstrato, que te permitiria de futuro usar outra coleção no `GameLoop` sem qualquer impacto no restante código?

PmaiWoW committed:
Update solution to ex06 of UML

Modify players collection in Map.cs from list to ICollection;
Remove 3 Map type variables from Game.cs, which now saves every
map created in the same variable (currentMap);
Remove addition of new Player type objects into an already existing
list, creating a new list, saving it into the players list;

nunofachada commented:
Porque razão instancias uma lista vazia logo no início do _game loop_ e logo a seguir gravas por cima uma nova lista já com jogadores? Nesse sítio não basta declarares a variável que contem a lista?

Penso que é a única coisa que falta neste exercício. Por favor altera, testa tudo, confirma que funciona e faz o _push_.

PmaiWoW committed:
Update solution to ex06 of UML

PmaiWoW commented:
Foi um lapso, não cheguei a retirar essa instanciação da lista antes de enviar, mas já foi alterado

nunofachada commented:
Testaste este código e funciona?

nunofachada commented:
Testaste este código e funciona?

PmaiWoW commented:
Sim, testei antes de enviar e funciona

nunofachada commented:
Ok Pedro, solução aceite.

PmaiWoW commented:
Mas houve algum problema do seu lado com o programa?

nunofachada commented:
Não, tudo OK.
